internal/controller/tempo/tempostack_controller.go

```diff
@@ -276,3 +313,83 @@ func (r *TempoStackReconciler) PatchStatus(ctx context.Context, changed, origina
 	statusPatch := client.MergeFrom(original)
 	return r.Client.Status().Patch(ctx, changed, statusPatch)
 }
+
+func finalize(ctx context.Context, cl client.Client, logger logr.Logger, objectLabels map[string]string) error {
```

Comment:
nit: because this function (and the new ones below) are shared between both controllers, I'd move them to a separate file